### PR TITLE
Openfga relation interface update 01

### PR DESCRIPTION
**Pull request by alesstimec**
## Description

The what and why - include a summary of the change, describe what it does, and include relevant motivation and context.

Fixes _JIRA/GitHub issue number_

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [ ] Covered by unit tests
- [ ] Covered by integration tests

### CODE REVIEW

File: charms/openfga-k8s/lib/charms/openfga_k8s/v0/openfga.py [hunk 1]
```diff
@@ -49,6 +49,11 @@ def _on_openfga_store_created(self, event: OpenFGAStoreCreateEvent):
         logger.info("address {}".format(event.address))
         logger.info("port {}".format(event.port))
         logger.info("scheme {}".format(event.scheme))
+
+        if event.token_secret_id:
+            secret = self.model.get_secret(id=event.token_secret_id)
+            content = secret.get_content()
+            logger.info("secret token {}".format(content["token"]))
```
Comment (kian99) [inline charms/openfga-k8s/lib/charms/openfga_k8s/v0/openfga.py:56]: Should we logging the secret content? Or will Juju hide it?

File: charms/openfga-k8s/src/charm.py [hunk 2]
```diff
@@ -140,42 +150,26 @@ def __init__(self, *args):
...
-        )
-        self.service_patcher = KubernetesServicePatch(
-            self, [port_http, port_grpc]
-        )
+        self.framework.observe(self.on.database_relation_broken, self._on_database_relation_broken)
 
-        self.state = PeerRelationState(self.model, self.app, "openfga-peer")
+        port_http = ServicePort(8080, name=f"{self.app.name}-http", protocol="TCP")
+        port_grpc = ServicePort(8081, name=f"{self.app.name}-grpc", protocol="TCP")
+        self.service_patcher = KubernetesServicePatch(self, [port_http, port_grpc])
```
Comment (kian99) [inline charms/openfga-k8s/src/charm.py:172]: I think Ali might have a way of handling multiple ingress ports without the K8ServicePatch now. Something for the future.

File: charms/openfga-k8s/lib/charms/openfga_k8s/v0/openfga.py (hunk 1 above)
Comment (babakks) [inline charms/openfga-k8s/lib/charms/openfga_k8s/v0/openfga.py:56]: Is it okay to log the token? In another place we're just logging the secret ID, but here it's the actual content.

File: charms/openfga-k8s/tox.ini [hunk 3]
```diff
@@ -61,7 +61,7 @@ deps =
     -r{toxinidir}/requirements.txt
 commands =
     coverage run --source={[vars]src_path} \
-        -m pytest --ignore={[vars]tst_path}integration -v --tb native -s {posargs}
+        -m pytest --ignore={[vars]tst_path}integration -vv --tb native -s {posargs}
```
Comment (babakks) [inline charms/openfga-k8s/tox.ini:64]: This `-vv` is different from the integration test environment below (there it's `-v`). Is it intentional?

**Review by babakks — APPROVED** [review]

File: charms/openfga-k8s/src/charm.py [hunk 4]
```diff
@@ -200,25 +194,33 @@ def _on_update_status(self, _):
...
         self._ready()
 
+    def _create_token_secret(self, event):
+        if not self._state.is_ready():
+            event.defer()
+            logger.warn("state not ready")
+            return
+
+        if not self._state.token_secret_id:
+            token = secrets.token_urlsafe(32)
```
Comment (kian99) [inline charms/openfga-k8s/src/charm.py:204]: What is this 32 bit? Should we make it a constant so it has a name?

File: charms/openfga-k8s/src/charm.py [hunk 5]
```diff
@@ -237,81 +239,67 @@ def _update_workload(self, event):
...
             return
 
+        if not self._state.is_ready():
+            logger.info("state is not ready")
+            event.defer()
+            return
+
+        self._create_token_secret(event)
+
+        # Quickly update logrotates config each workload update
```
Comment (kian99) [inline charms/openfga-k8s/src/charm.py:249]: It's python so arguably it won't be running quickly.

File: charms/openfga-k8s/src/charm.py [hunk 6]
```diff
@@ -606,171 +590,157 @@ def _on_schema_upgrade_action(self, event):
...
 
-                new_csr = generate_csr(
-                    private_key=private_key.encode(),
-                    subject=dnsname,
-                )
-                self.certificates.request_certificate_renewal(
-                    old_certificate_signing_request=old_csr,
-                    new_certificate_signing_request=new_csr,
-                )
+        if not self._state.is_ready():
```
Comment (kian99) [inline charms/openfga-k8s/src/charm.py:680]: We could probably just make a decorator called "requires_state" that we add to all these functions.

File: charms/openfga-k8s/src/charm.py [hunk 7]
```diff
@@ -606,171 +590,157 @@ def _on_schema_upgrade_action(self, event):
...
+            logger.info("state not ready")
+            event.defer()
+            return
+
+        del self._state.dns_name
 
         self._update_workload(event)
 
+<<<<<<< HEAD
```
Comment (kian99) [inline charms/openfga-k8s/src/charm.py:739]: Some git conflict stuff here.

File: charms/openfga-k8s/tests/unit/test_charm.py [hunk 8]
```diff
@@ -179,7 +153,20 @@ def test_on_openfga_relation_joined(
...
             "port": "8080",
             "scheme": "http",
-            "token": "test-token",
+            "token_secret_id": test_regex("secret:.*"),
             "store_id": "01GK13VYZK62Q1T0X55Q2BHYD6",
             "dns_name": "openfga-k8s-0.openfga-k8s-endpoints.None.svc.cluster.local",
         }
+
+
+class test_regex:
```
Comment (kian99) [inline charms/openfga-k8s/tests/unit/test_charm.py:162]: Is this needed?

**Review by kian99 — APPROVED** [review]
Some things to address but otherwise looks good.

File: charms/openfga-k8s/tests/unit/test_charm.py (hunk 8 above)
Comment (alesstimec) [reply charms/openfga-k8s/tests/unit/test_charm.py:162]: what is the best way to assert string matches a regexp?

File: charms/openfga-k8s/tox.ini (hunk 3 above)
Comment (alesstimec) [reply charms/openfga-k8s/tox.ini]: reverted

File: charms/openfga-k8s/src/charm.py (hunk 4 above)
Comment (alesstimec) [reply charms/openfga-k8s/src/charm.py:180]: i don't know.. it think this is ok

File: charms/openfga-k8s/src/charm.py (hunk 7 above)
Comment (alesstimec) [reply charms/openfga-k8s/src/charm.py]: resolved

File: charms/openfga-k8s/lib/charms/openfga_k8s/v0/openfga.py (hunk 1 above)
Comment (alesstimec) [reply charms/openfga-k8s/lib/charms/openfga_k8s/v0/openfga.py]: changed..

**Review by babakks — APPROVED** [review]

File: charms/openfga-k8s/metadata.yaml [hunk 9]
```diff
@@ -37,8 +37,8 @@ description: |
   Community section for third-party SDKs and tools.
 
 peers:
-  openfga-peer:
-    interface: openfga-peer
+  peer:
+    interface: openfga_peer
```
Comment (kian99) [inline charms/openfga-k8s/metadata.yaml:41]: To be fair I do think the `-` gets converted to `_` for use in any function signatures.

File: charms/openfga-k8s/src/charm.py [hunk 10]
```diff
@@ -200,35 +186,37 @@ def _on_update_status(self, _):
...
+    def _create_token_secret(self, event):
+        if not self._state.token_secret_id:
+            token = secrets.token_urlsafe(32)
+            content = {"token": token}
+            secret = self.app.add_secret(content)
+            self._state.token_secret_id = secret.id
+            logger.info("created token secret {}".format(secret.id))
+
+    @must_be_leader
+    @requires_state
```
Comment (kian99) [inline charms/openfga-k8s/src/charm.py:199]: Really like the decorators, very nice.

File: charms/openfga-k8s/tests/unit/test_charm.py (hunk 8 above)
Comment (kian99) [reply charms/openfga-k8s/tests/unit/test_charm.py:160]: Just curious if the class is needed, instead of just using the functions in this class directly.

**Review by kian99 — APPROVED** [review]

**Comment by kian99** [comment]
Oh and gentle reminder about committing without `--amend --no-edit` sometimes :D

**Pull request merged**